frontend/src/metabase/setup/components/WelcomePage/WelcomePage.tsx

```diff
@@ -11,26 +12,34 @@ import {
 } from "./WelcomePage.styled";
 
 export interface WelcomePageProps {
+  isLocaleLoaded: boolean;
   onStepShow: () => void;
   onStepSubmit: () => void;
 }
 
 const WelcomePage = ({
+  isLocaleLoaded,
   onStepShow,
   onStepSubmit,
-}: WelcomePageProps): JSX.Element => {
+}: WelcomePageProps): JSX.Element | null => {
+  const isElapsed = useIsElapsed(LOCALE_TIMEOUT);
+
   useEffect(() => {
     onStepShow();
   }, [onStepShow]);
 
+  if (!isElapsed && !isLocaleLoaded) {
```

Comment:
In case locale data doesn't load in 300 ms, show the user interface in English. By doing that we:
- prevent a "flash" when immediately changing texts from English to the default language after initial render
- handle locale loading issues, so in case it doesn't load we will show the UI in English and update it when the data is loaded